access/src/main/external/static/js/admin/src/MoveActionMonitor.js

```diff
@@ -0,0 +1,250 @@
+
+		for (var i = A.length; i--; ) {
+			if (!map.hasOwnProperty(A[i]))
+				C.push(A[i]);
+		}
+
+		return C;
+	}
+	
+	MoveActionMonitor.prototype.setResultList = function(resultList) {
+		this.resultList = resultList;
+	};
+	
+	MoveActionMonitor.prototype.activate = function() {
+		this.active = true;
+		
+		// Mark objects that were previously known to be moving
+		var self = this;
+		$.each(this.moveData, function(moveId, moveDetails) {
+			self.markMoving(moveDetails.pids);
+		});
+		
+		// Start refresh loop
+		this.update();
+	};
+	
+	MoveActionMonitor.prototype.update = function() {
+		var self = this;
+		
+		$.ajax({
+			url : "/services/api/listMoves",
+			contentType: "application/json; charset=utf-8",
+			dataType: "json"
+		}).done(function(data) {
+			var remoteMoves = data == null? {} : data;
+			
+			try {
+				// Clean up inactive move operations that have been removed remotely
+				self.cleanMoveTombstones(remoteMoves);
+				// Clean up completed move operations
+				self.completeMoves(remoteMoves);
+			
+				// Indicate new move options
+				self.addMoves(remoteMoves);
+				
+			} finally {
+				// Queue up the next run
+				setTimeout($.proxy(self.update, self), self.options.updateInterval);
+			}
+		}).fail(function() {
+			console.error("Failed to retrieve move information from server");
+			// Keep refreshing alive, but wait a bit longer
+			setTimeout($.proxy(self.update, self), self.options.updateInterval * 2);
+		});
+	};
+	
+	MoveActionMonitor.prototype.cleanMoveTombstones = function(remoteMoves) {
+
+		var localInactive = [];
+		for (var moveId in this.moveData) {
+			if (this.moveData[moveId].inactive) {
+				localInactive.push(moveId);
+			}
+		}
+		
+		if (localInactive.length == 0)
+			return;
+
+		// Cleanable tombstones = Inactive Local Moves - Remote Moves
+		var cleanable = setDifference(localInactive, Object.keys(remoteMoves));
+		
+		if (cleanable.length == 0)
+			return;
+		
+		for (var index in cleanable) {
+			var cleanableId = cleanable[index];
+			delete this.moveData[cleanableId];
+			this.removeUserMove(cleanableId);
+		}
+	};
+	
+	MoveActionMonitor.prototype.completeMoves = function(remoteMoves) {
+
+		// Get the set of move ids which were in progress according to the server
+		var remoteActive = [];
+		for (var moveId in remoteMoves) {
+			if (remoteMoves[moveId]) {
+				remoteActive.push(moveId);
+			}
+		}
+		
+		// Get the list of move ids considered in progress locally
+		var localActive = [];
+		for (var moveId in this.moveData) {
+			if (!this.moveData[moveId].inactive) {
+				localActive.push(moveId);
+			}
+		}
+		
+		var localMoves = Object.keys(this.moveData);
+		// Completed jobs = Local Moves - Active Remote Moves
+		var completed = setDifference(localActive, remoteActive);
+		
+		if (completed.length == 0)
+			return;
+		
+		var userMoves = this.getUserMoves();
+		// Process the set of completed moves to remove them from local memory and inform the user
+		for (var index in completed) {
```

Comment:
Which structure would you prefer?  It seems about the same as a for (var i = 0; i < completed.length; i++) and I generally haven't been using .each